### PR TITLE
feat: read/write files for refg name hash lookups

### LINKED ISSUE #60: Reverse-lookup refg name from hash
Refg names are not encoded as part of intervals. Instead, it's a 32-bit hash of the name.
There is no way for GK to know the refg name from the hash. If a user loads a GK binary file (e.g gtrack) and tries to get the refg name from an interval, they get a cryptic error:
```
ValueError: src/refg.cpp:74: Could not retrieve name for 18068147913528327027
```

### PR DESCRIPTION
**Pull request by ovesh**
fixes #60

### CODE REVIEW

File: genome_kit/gk_data.py [hunk 1]
```diff
@@ -56,6 +56,17 @@ def get_file(filename):
 def upload_file(filepath:str, filename:str, metadata:Dict[str, str]=None):  # pragma: no cover
     """compatibility wrapper for :py:meth:`~genome_kit.DataManager.upload_file`"""
 
+    chrom_sizes_ext = ".chrom.sizes"
+    if filename.endswith(chrom_sizes_ext):
+        refg_name = filename[:-len(chrom_sizes_ext)]
+        hashval = _cxx.Genome._refg_hash(refg_name)
```
Comment (s22chan) [inline genome_kit/gk_data.py:62]: Can probably just do Genome._refg_hash but either is fine

File: src/refg.cpp [hunk 2]
```diff
@@ -68,9 +69,52 @@ refg_t refg_registry_t::as_refg(std::string_view config) const
...
+	std::string path{fmt::format("{}.hash", ref)};
+	if (!std::filesystem::exists(path)) {
+		path = resolve_datafile_path(prepend_dir(data_dir(), path));
+	}
+	if (!std::filesystem::exists(path)) {
+		return {};
+	}
+	std::FILE* file = std::fopen(path.c_str(), "rb");
+	if (!file) {
+		println("Could not open file '{}'", path);
```
Comment (s22chan) [inline src/refg.cpp:83]: should print the strerr as well. although I think maybe just using the line reader would be less reinvention

File: src/refg.cpp [hunk 3]
```diff
@@ -68,9 +69,52 @@ refg_t refg_registry_t::as_refg(std::string_view config) const
...
+	auto bytesRead = std::fread(&refg_name[0], 1, max_size, file);
+	if (std::ferror(file)) {
+        std::fclose(file);
+		println("Could not read file '{}'", path);
+		return {};
+    }
+	std::fclose(file);
+	refg_name.resize(bytesRead);
+	refg_name.erase(refg_name.find_last_not_of(" \n\r\t") + 1);
+	auto expected_ref = fnv1a_hash64(std::string_view(refg_name));
```
Comment (s22chan) [inline src/refg.cpp:98]: don't need to cast to string_view

File: src/refg.cpp [hunk 4]
```diff
@@ -68,9 +69,52 @@ refg_t refg_registry_t::as_refg(std::string_view config) const
...
+	GK_CHECK(name_inserted || name_it->second == refg_name, runtime,
+			 "hash collision, try renaming one of the assemblies: '{}' and '{}'", name_it->first, refg_name);
+
+	return name_it->second;
+}
+
 std::string_view refg_registry_t::refg_as_sv(refg_t ref) const
 {
 	const auto it = _names_by_refg.find(ref);
+	if (it == std::end(_names_by_refg)) {
```
Comment (s22chan) [inline src/refg.cpp:112]: doing a double lookup here. could do a lower_bound and an insertion hint to fix this but the amount of genomes loaded is probably very low

File: genome_kit/gk_data.py [hunk 5]
```diff
@@ -56,6 +56,17 @@ def get_file(filename):
 def upload_file(filepath:str, filename:str, metadata:Dict[str, str]=None):  # pragma: no cover
     """compatibility wrapper for :py:meth:`~genome_kit.DataManager.upload_file`"""
 
+    chrom_sizes_ext = ".chrom.sizes"
```
Comment (s22chan) [inline genome_kit/gk_data.py:59]: I know I was the one that recommended this, but is there a way we can support offline users as well (for instance, that uses the starter script)

File: src/refg.cpp [hunk 6]
```diff
@@ -68,9 +69,52 @@ refg_t refg_registry_t::as_refg(std::string_view config) const
 	return ref;
 }
 
+std::string_view refg_registry_t::_try_refg_as_sv_from_file(refg_t ref) const
```
Comment (s22chan) [inline src/refg.cpp:72]: would be good to have a test

File: src/refg.cpp [hunk 7]
```diff
@@ -68,9 +69,52 @@ refg_t refg_registry_t::as_refg(std::string_view config) const
...
+	if (std::ferror(file)) {
+        std::fclose(file);
+		println("Could not read file '{}'", path);
+		return {};
+    }
+	std::fclose(file);
+	refg_name.resize(bytesRead);
+	refg_name.erase(refg_name.find_last_not_of(" \n\r\t") + 1);
+	auto expected_ref = fnv1a_hash64(std::string_view(refg_name));
+	GK_CHECK(expected_ref == (uint64_t)ref, runtime, "Hash mismatch in '{}' for '{}': {} != {}",
```
Comment (s22chan) [inline src/refg.cpp:99]: this doesn't leak the underlying type

```suggestion
	GK_CHECK(refg_t(expected_ref) == ref, runtime, "Hash mismatch in '{}' for '{}': {} != {}",
```

File: src/refg.cpp [hunk 8]
```diff
@@ -68,9 +69,52 @@ refg_t refg_registry_t::as_refg(std::string_view config) const
...
 }
 
+std::string_view refg_registry_t::_try_refg_as_sv_from_file(refg_t ref) const
+{
+	std::string path{fmt::format("{}.hash", ref)};
+	if (!std::filesystem::exists(path)) {
+		path = resolve_datafile_path(prepend_dir(data_dir(), path));
+	}
+	if (!std::filesystem::exists(path)) {
+		return {};
```
Comment (s22chan) [inline src/refg.cpp:79]: this is returning pointers to local variables (that will be delteted).

change the fn signature to return a `std::string` instead

File: src/refg.cpp [hunk 9]
```diff
@@ -68,9 +69,52 @@ refg_t refg_registry_t::as_refg(std::string_view config) const
...
+    }
+	std::fclose(file);
+	refg_name.resize(bytesRead);
+	refg_name.erase(refg_name.find_last_not_of(" \n\r\t") + 1);
+	auto expected_ref = fnv1a_hash64(std::string_view(refg_name));
+	GK_CHECK(expected_ref == (uint64_t)ref, runtime, "Hash mismatch in '{}' for '{}': {} != {}",
+			 path, refg_name, expected_ref, ref);
+
+	const auto [name_it, name_inserted] = _names_by_refg.try_emplace(ref, refg_name);
```
Comment (s22chan) [inline src/refg.cpp:102]: this should probably go in the caller level to separate responsibilities cleanly.

(eg IO in this function, memory map in the other)

**Review by s22chan — CHANGES_REQUESTED** [review]

File: genome_kit/gk_data.py (hunk 1 above)
Comment (ovesh) [reply genome_kit/gk_data.py:62]: If I `from . import Genome` I get a circular dependency.

File: src/refg.cpp (hunk 8 above)
Comment (s22chan) [reply src/refg.cpp:79]: actually I misread this (long day).  default ctor on a string_view is fine since it doesn't have a ptr to anything.

**Review by s22chan — APPROVED** [review]

File: src/refg.cpp (hunk 2 above)
Comment (ovesh) [reply src/refg.cpp]: Using line_reader

File: src/refg.cpp (hunk 4 above)
Comment (ovesh) [reply src/refg.cpp:99]: Meh, I think we can take the very small performance hit. Thanks for the tip anyway.

File: genome_kit/gk_data.py (hunk 5 above)
Comment (ovesh) [reply genome_kit/gk_data.py:59]: I'll add hash files to the starter script.

File: src/refg.cpp [hunk 10]
```diff
@@ -68,9 +69,39 @@ refg_t refg_registry_t::as_refg(std::string_view config) const
...
+	if (!std::filesystem::exists(path)) {
+		path = resolve_datafile_path(prepend_dir(data_dir(), path));
+	}
+	if (!std::filesystem::exists(path)) {
+		return {};
+	}
+
+	line_reader lr{path};
+	auto refg_name = std::string(lr.line());
+	refg_name.erase(refg_name.find_last_not_of(" \n\r\t") + 1);
```
Comment (s22chan) [inline src/refg.cpp:84]:
```suggestion
    auto refg_name = strip(lr.line());
```

File: src/refg.cpp (hunk 6 above)
Comment (ovesh) [reply src/refg.cpp:72]: I'm not sure how to implement a test w/o first loading the relevant Genome (thus making the test kinda meaningless).

Best I can come up with is to add a small binary file (e.g gtrack) to the repo, but I'd rather not.

File: src/refg.cpp [hunk 11]
```diff
@@ -98,6 +94,9 @@ std::string_view refg_registry_t::refg_as_sv(refg_t ref) const
 	if (it == std::end(_names_by_refg)) {
 		const auto& refg_name = _try_refg_as_sv_from_file(ref);
 		if (!refg_name.empty()) {
+			const auto [name_it, name_inserted] = _names_by_refg.try_emplace(ref, refg_name);
+			GK_CHECK(name_inserted || name_it->second == refg_name, runtime,
+					 "hash collision, try renaming one of the assemblies: '{}' and '{}'", name_it->first, refg_name);
 			return refg_name;
```
Comment (s22chan) [inline src/refg.cpp:100]:
```suggestion
	if (!refg_name.empty()) {
			const auto [it, name_inserted] = _names_by_refg.try_emplace(ref, refg_name);
			GK_CHECK(name_inserted || it->second == refg_name, runtime,
					 "hash collision, try renaming one of the assemblies: '{}' and '{}'", it->second, refg_name);
```

**Pull request merged**